File: agents/skills/manus/scripts/manus_client.mjs
/* global fetch, process, URL */

import { createWriteStream, existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { basename, dirname, join } from "node:path";
import { Readable } from "node:stream";
import { finished } from "node:stream/promises";

const BASE_URL = "https://api.manus.ai";
const STATE_HOME = process.env.MANUS_SKILL_HOME || join(homedir(), ".manus-skill");
const REGISTRY_PATH = join(STATE_HOME, "cache", "tasks.json");
const DEFAULT_MEDIA_DIR = join(STATE_HOME, "downloads");

function fail(message, code = 1) {
  process.stderr.write(`${JSON.stringify({ error: message })}\n`);
  process.exit(code);
}

function getApiKey() {
  const key = process.env.MANUS_API_KEY || "";
  if (!key) {
    fail("MANUS_API_KEY not set");
  }
  return key;
}

function apiHeaders(extra = {}) {
  return {
    API_KEY: getApiKey(),
    "Content-Type": "application/json",
    ...extra,
  };
}

function monthDir() {
  const now = new Date();
  const y = now.getUTCFullYear();
  const m = String(now.getUTCMonth() + 1).padStart(2, "0");
  const dir = join(DEFAULT_MEDIA_DIR, `${y}${m}`);
  mkdirSync(dir, { recursive: true });
  return dir;
}

function loadRegistry() {
  if (!existsSync(REGISTRY_PATH)) {
    return {};
  }
  try {
    return JSON.parse(readFileSync(REGISTRY_PATH, "utf8"));
  } catch {
    return {};
  }
}

function saveRegistry(data) {
  mkdirSync(join(STATE_HOME, "cache"), { recursive: true });
  writeFileSync(REGISTRY_PATH, `${JSON.stringify(data, null, 2)}\n`, "utf8");
}

function registerTask(taskId, label, promptSummary) {
  const registry = loadRegistry();
  registry[taskId] = {
    label: label || "",
    created_at: new Date().toISOString(),
    prompt_summary: promptSummary.slice(0, 200),
    status: "created",
  };
  saveRegistry(registry);
}

function updateTaskStatus(taskId, status) {
  const registry = loadRegistry();
  if (!registry[taskId]) {
    return;
  }
  registry[taskId].status = status;
  saveRegistry(registry);
}

function safeOutputPath(downloadDir, filename) {
  let cleaned = basename(filename || "attachment");
  if (!cleaned || cleaned === "." || cleaned === "..") {
    cleaned = "attachment";
  }

  const dotIndex = cleaned.lastIndexOf(".");
  const stem = dotIndex > 0 ? cleaned.slice(0, dotIndex) : cleaned;
  const suffix = dotIndex > 0 ? cleaned.slice(dotIndex) : "";

  let candidate = join(downloadDir, cleaned);
  let index = 1;
  while (existsSync(candidate)) {
    candidate = join(downloadDir, `${stem}-${index}${suffix}`);
    index += 1;
  }
  return candidate;
}

async function requestJson(url, init = {}) {
  const response = await fetch(url, init);
  if (!response.ok) {
    const body = await response.text();
    fail(`HTTP ${response.status} for ${url}: ${body}`);
  }
  return response.json();
}

async function uploadFile(filepath) {
  if (!existsSync(filepath)) {
    fail(`File not found: ${filepath}`);
  }

  const fileInfo = await requestJson(`${BASE_URL}/v1/files`, {
    method: "POST",
    headers: apiHeaders(),
    body: JSON.stringify({ filename: basename(filepath) }),
  });

  const fileBuffer = readFileSync(filepath);
  const putResponse = await fetch(fileInfo.upload_url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/octet-stream",
      "Content-Length": String(fileBuffer.byteLength),
    },
    body: fileBuffer,
  });
  if (!putResponse.ok) {
    const body = await putResponse.text();
    fail(`Upload failed for ${filepath}: HTTP ${putResponse.status}: ${body}`);
  }

  return {
    filename: basename(filepath),
    file_id: fileInfo.id,
  };
}
async function downloadFile(url, destination) {
  const response = await fetch(url);
  if (!response.ok || !response.body) {
    const body = await response.text();
    fail(`Download failed for ${url}: HTTP ${response.status}: ${body}`);
  }

  mkdirSync(dirname(destination), { recursive: true });
  const stream = createWriteStream(destination);
  await finished(Readable.fromWeb(response.body).pipe(stream));
}

function printJson(data) {
  process.stdout.write(`${JSON.stringify(data, null, 2)}\n`);
}

function printMedia(path) {
  process.stderr.write(`MEDIA:${path}\n`);
}

export function collectAssistantOutputs(data) {
  const texts = [];
  const files = [];

  for (const entry of data.output || []) {
    if (entry.role !== "assistant") {
      continue;
    }

    for (const content of entry.content || []) {
      if (content.text) {
        texts.push(content.text);
      }
      if (content.fileUrl) {
        files.push({
          url: content.fileUrl,
          name: content.fileName || "attachment",
        });
      }
    }
  }

  return { texts, files };
}

function parseArgs(argv) {
  const [command, ...rest] = argv;
  if (!command) {
    fail("Missing command");
  }

  const options = {
    command,
    attachment: [],
    connector: [],
    convert: false,
    interactive: false,
    mode: "agent",
    profile: "manus-1.6",
  };

  for (let i = 0; i < rest.length; i += 1) {
    const arg = rest[i];
    switch (arg) {
      case "--prompt":
        options.prompt = rest[++i];
        break;
      case "--mode":
        options.mode = rest[++i];
        break;
      case "--profile":
        options.profile = rest[++i];
        break;
      case "--label":
        options.label = rest[++i];
        break;
      case "--attachment":
        options.attachment.push(rest[++i]);
        break;
      case "--connector":
        options.connector.push(rest[++i]);
        break;
      case "--locale":
        options.locale = rest[++i];
        break;
      case "--task-id":
        options.taskId = rest[++i];
        break;
      case "--download-dir":
        options.downloadDir = rest[++i];
        break;
      case "--limit":
        options.limit = Number.parseInt(rest[++i], 10);
        break;
      case "--status":
        options.status = rest[++i];
        break;
      case "--convert":
        options.convert = true;
        break;
      case "--interactive":
        options.interactive = true;
        break;
      default:
        fail(`Unknown argument: ${arg}`);
    }
  }

  return options;
}

async function cmdCreate(args) {
  if (!args.prompt) {
    fail("--prompt is required");
  }

  const payload = {
    prompt: args.prompt,
    agentProfile: args.profile,
  };
  if (args.mode) payload.taskMode = args.mode;
  if (args.locale) payload.locale = args.locale;
  if (args.taskId) payload.taskId = args.taskId;
  if (args.interactive) payload.interactiveMode = true;
  if (args.attachment.length > 0) {
    payload.attachments = [];
    for (const filepath of args.attachment) {
      payload.attachments.push(await uploadFile(filepath));
    }
  }
  if (args.connector.length > 0) {
    payload.connectors = args.connector;
  }

  const result = await requestJson(`${BASE_URL}/v1/tasks`, {
    method: "POST",
    headers: apiHeaders(),
    body: JSON.stringify(payload),
  });

  if (result.task_id) {
    registerTask(result.task_id, args.label, args.prompt);
  }
  printJson(result);
}

async function getTask(taskId, convert) {
  if (!taskId) {
    fail("--task-id is required");
  }
  const url = new URL(`${BASE_URL}/v1/tasks/${taskId}`);
  if (convert) {
    url.searchParams.set("convert", "true");
  }
  return requestJson(url, {
    method: "GET",
    headers: apiHeaders(),
  });
}

async function cmdStatus(args) {
  const data = await getTask(args.taskId, args.convert);
  const status = data.status || "unknown";
  updateTaskStatus(args.taskId, status);
  const summary = {
    task_id: args.taskId,
    status,
    created_at: data.created_at,
    updated_at: data.updated_at,
    credit_usage: data.credit_usage,
  };
  if (data.error) {
    summary.error = data.error;
  }
  printJson(summary);
}

async function cmdResult(args) {
  const data = await getTask(args.taskId, args.convert);
  const status = data.status || "unknown";
  updateTaskStatus(args.taskId, status);

  const downloadDir = args.downloadDir || monthDir();
  mkdirSync(downloadDir, { recursive: true });

  const downloadedFiles = [];
  const { texts, files } = collectAssistantOutputs(data);
  for (const file of files) {
    const destination = safeOutputPath(downloadDir, file.name);
    await downloadFile(file.url, destination);
    downloadedFiles.push(destination);
  }

  printJson({
    task_id: args.taskId,
    status,
    text: texts.length > 0 ? texts.join("\n\n") : null,
    files: downloadedFiles,
    credit_usage: data.credit_usage,
  });

  for (const file of downloadedFiles) {
    printMedia(file);
  }
}

async function cmdList(args) {
  const url = new URL(`${BASE_URL}/v1/tasks`);
  url.searchParams.set("limit", String(args.limit || 10));
  url.searchParams.set("order", "desc");
  if (args.status) {
    url.searchParams.append("status", args.status);
  }

  const data = await requestJson(url, {
    method: "GET",
    headers: apiHeaders(),
  });

  const tasks = (data.data || []).map((task) => ({
    task_id: task.id,
    status: task.status,
    created_at: task.created_at,
    credit_usage: task.credit_usage,
  }));

  printJson({
    tasks,
    has_more: Boolean(data.has_more),
  });
}

async function cmdDelete(args) {
  if (!args.taskId) {
    fail("--task-id is required");
  }
  const data = await requestJson(`${BASE_URL}/v1/tasks/${args.taskId}`, {
    method: "DELETE",
    headers: { API_KEY: getApiKey() },
  });
  updateTaskStatus(args.taskId, "deleted");
  printJson(data);
}

export async function main() {
  const args = parseArgs(process.argv.slice(2));
  switch (args.command) {
    case "create":
      await cmdCreate(args);
      break;
    case "status":
      await cmdStatus(args);
      break;
    case "result":
      await cmdResult(args);
      break;
    case "list":
      await cmdList(args);
      break;
    case "delete":
    case "cancel":
      await cmdDelete(args);
      break;
    default:
      fail(`Unknown command: ${args.command}`);
  }
}

if (import.meta.main) {
  await main();
}

File: agents/skills/manus/scripts/manus_client.py
# /// script
# requires-python = ">=3.11"
# dependencies = ["httpx>=0.27.0"]
# ///
"""Generic Manus API client.

Usage:
    uv run manus_client.py create --prompt "..." [options]
    uv run manus_client.py status --task-id <id>
    uv run manus_client.py result --task-id <id> [--download-dir <path>] [--convert]
    uv run manus_client.py list [--limit N] [--status <status>]
    uv run manus_client.py delete --task-id <id>
"""

from __future__ import annotations

import argparse
import json
import os
import sys
from datetime import datetime, timezone
from json import JSONDecodeError
from pathlib import Path

import httpx

BASE_URL = "https://api.manus.ai"
STATE_HOME = Path(os.environ.get("MANUS_SKILL_HOME", Path.home() / ".manus-skill"))
REGISTRY_PATH = STATE_HOME / "cache" / "tasks.json"
DEFAULT_MEDIA_DIR = STATE_HOME / "downloads"


def get_api_key() -> str:
    key = os.environ.get("MANUS_API_KEY", "")
    if not key:
        print('{"error": "MANUS_API_KEY not set"}', file=sys.stderr)
        sys.exit(1)
    return key


def api_headers() -> dict[str, str]:
    return {
        "API_KEY": get_api_key(),
        "Content-Type": "application/json",
    }


def media_dir_for_now() -> Path:
    now = datetime.now(timezone.utc)
    d = DEFAULT_MEDIA_DIR / now.strftime("%Y%m")
    d.mkdir(parents=True, exist_ok=True)
    return d


def safe_output_path(download_dir: Path, filename: str) -> Path:
    cleaned = Path(filename or "attachment").name
    if not cleaned or cleaned in {".", ".."}:
        cleaned = "attachment"

    candidate = download_dir / cleaned
    stem = candidate.stem or "attachment"
    suffix = candidate.suffix
    index = 1
    while candidate.exists():
        candidate = download_dir / f"{stem}-{index}{suffix}"
        index += 1
    return candidate


def collect_assistant_outputs(data: dict) -> tuple[list[str], list[dict[str, str]]]:
    texts: list[str] = []
    files: list[dict[str, str]] = []

    for entry in data.get("output", []):
        if entry.get("role") != "assistant":
            continue

        for content in entry.get("content", []):
            if content.get("text"):
                texts.append(content["text"])
            if content.get("fileUrl"):
                files.append(
                    {
                        "url": content["fileUrl"],
                        "name": content.get("fileName", "attachment"),
                    }
                )

    return texts, files


# --- Task Registry ---

def load_registry() -> dict:
    if REGISTRY_PATH.exists():
        try:
            return json.loads(REGISTRY_PATH.read_text())
        except (OSError, JSONDecodeError):
            return {}
    return {}


def save_registry(data: dict) -> None:
    REGISTRY_PATH.parent.mkdir(parents=True, exist_ok=True)
    REGISTRY_PATH.write_text(f"{json.dumps(data, indent=2, ensure_ascii=False)}\n")


def register_task(task_id: str, label: str | None, prompt_summary: str) -> None:
    reg = load_registry()
    reg[task_id] = {
        "label": label or "",
        "created_at": datetime.now(timezone.utc).isoformat(),
        "prompt_summary": prompt_summary[:200],
        "status": "created",
    }
    save_registry(reg)


def update_task_status(task_id: str, status: str) -> None:
    reg = load_registry()
    if task_id in reg:
        reg[task_id]["status"] = status
        save_registry(reg)


# --- File Upload ---

def upload_file(filepath: str) -> dict[str, str]:
    """Upload a file via presigned URL, return attachment payload."""
    p = Path(filepath)
    if not p.exists():
        print(json.dumps({"error": f"File not found: {filepath}"}), file=sys.stderr)
        sys.exit(1)

    with httpx.Client(timeout=60) as client:
        # Step 1: Get presigned URL
        resp = client.post(
            f"{BASE_URL}/v1/files",
            headers=api_headers(),
            json={"filename": p.name},
        )
        resp.raise_for_status()
        file_info = resp.json()
        file_id = file_info["id"]
        upload_url = file_info["upload_url"]

        # Step 2: PUT file to presigned URL
        with open(p, "rb") as f:
            put_resp = client.put(
                upload_url,
                content=f,
                headers={"Content-Type": "application/octet-stream"},
            )
            put_resp.raise_for_status()

    return {
        "filename": p.name,
        "file_id": file_id,
    }


# --- Subcommands ---

def cmd_create(args: argparse.Namespace) -> None:
    payload: dict = {
        "prompt": args.prompt,
        "agentProfile": args.profile,
    }
    if args.mode:
        payload["taskMode"] = args.mode
    if args.locale:
        payload["locale"] = args.locale
    if args.task_id:
        payload["taskId"] = args.task_id
    if args.interactive:
        payload["interactiveMode"] = True

    # Handle file attachments
    if args.attachment:
        attachments = []
        for fpath in args.attachment:
            attachments.append(upload_file(fpath))
        payload["attachments"] = attachments

    # Handle connectors
    if args.connector:
        payload["connectors"] = args.connector

    with httpx.Client(timeout=120) as client:
        resp = client.post(
            f"{BASE_URL}/v1/tasks",
            headers=api_headers(),
            json=payload,
        )
        resp.raise_for_status()
        result = resp.json()

    task_id = result.get("task_id", "")

    if task_id:
        register_task(task_id, args.label, args.prompt)

    print(json.dumps(result, indent=2, ensure_ascii=False))


def cmd_status(args: argparse.Namespace) -> None:
    params = {"convert": "true"} if args.convert else None
    with httpx.Client(timeout=30) as client:
        resp = client.get(
            f"{BASE_URL}/v1/tasks/{args.task_id}",
            headers=api_headers(),
            params=params,
        )
        resp.raise_for_status()
        data = resp.json()

    status = data.get("status", "unknown")
    update_task_status(args.task_id, status)

    # Return a concise status summary
    summary = {
        "task_id": args.task_id,
        "status": status,
        "created_at": data.get("created_at"),
        "updated_at": data.get("updated_at"),
        "credit_usage": data.get("credit_usage"),
    }
    if data.get("error"):
        summary["error"] = data["error"]

    print(json.dumps(summary, indent=2, ensure_ascii=False))


def cmd_result(args: argparse.Namespace) -> None:
    params = {"convert": "true"} if args.convert else None
    with httpx.Client(timeout=60) as client:
        resp = client.get(
            f"{BASE_URL}/v1/tasks/{args.task_id}",
            headers=api_headers(),
            params=params,
        )
        resp.raise_for_status()
        data = resp.json()

    status = data.get("status", "unknown")
    update_task_status(args.task_id, status)

    download_dir = Path(args.download_dir) if args.download_dir else media_dir_for_now()
    download_dir.mkdir(parents=True, exist_ok=True)

    downloaded_files: list[str] = []
    texts, files = collect_assistant_outputs(data)

    for file_info in files:
        dest = safe_output_path(download_dir, file_info["name"])
        # Download assistant-emitted files only. User-uploaded input attachments
        # appear in task output too, but they are not task results.
        with httpx.Client(timeout=120) as dl_client, dl_client.stream(
            "GET",
            file_info["url"],
        ) as dl_resp:
            dl_resp.raise_for_status()
            with open(dest, "wb") as out:
                for chunk in dl_resp.iter_bytes():
                    out.write(chunk)
        downloaded_files.append(str(dest))

    result = {
        "task_id": args.task_id,
        "status": status,
        "text": "\n\n".join(texts) if texts else None,
        "files": downloaded_files,
        "credit_usage": data.get("credit_usage"),
    }

    print(json.dumps(result, indent=2, ensure_ascii=False))

    # Emit file paths on stderr for callers that want to inspect downloads.
    for fpath in downloaded_files:
        print(f"MEDIA:{fpath}", file=sys.stderr)


def cmd_list(args: argparse.Namespace) -> None:
    params: dict = {"limit": args.limit, "order": "desc"}
    if args.status:
        params["status"] = args.status

    with httpx.Client(timeout=30) as client:
        resp = client.get(
            f"{BASE_URL}/v1/tasks",
            headers=api_headers(),
            params=params,
        )
        resp.raise_for_status()
        data = resp.json()

    tasks = []
    for t in data.get("data", []):
        tasks.append({
            "task_id": t.get("id"),
            "status": t.get("status"),
            "created_at": t.get("created_at"),
            "credit_usage": t.get("credit_usage"),
        })

    print(json.dumps({"tasks": tasks, "has_more": data.get("has_more", False)}, indent=2, ensure_ascii=False))


def cmd_delete(args: argparse.Namespace) -> None:
    with httpx.Client(timeout=30) as client:
        resp = client.delete(
            f"{BASE_URL}/v1/tasks/{args.task_id}",
            headers={"API_KEY": get_api_key()},
        )
        resp.raise_for_status()
        data = resp.json()

    update_task_status(args.task_id, "deleted")
    print(json.dumps(data, indent=2, ensure_ascii=False))


def main() -> None:
    parser = argparse.ArgumentParser(description="Generic Manus API client")
    sub = parser.add_subparsers(dest="command", required=True)

    # create
    p_create = sub.add_parser("create", help="Create a Manus task")
    p_create.add_argument("--prompt", required=True, help="Task prompt")
    p_create.add_argument("--mode", choices=["agent", "adaptive", "chat"], default="agent")
    p_create.add_argument("--profile", default="manus-1.6", choices=["manus-1.6", "manus-1.6-lite", "manus-1.6-max"])
    p_create.add_argument("--label", default=None, help="Optional local label stored in the task registry")
    p_create.add_argument("--attachment", action="append", help="File path to attach (repeatable)")
    p_create.add_argument("--connector", action="append", help="Connector UUID (repeatable)")
    p_create.add_argument("--locale", default=None, help="Locale (e.g., zh-CN, en-US)")
    p_create.add_argument("--task-id", default=None, help="Existing task ID for multi-turn continuation")
    p_create.add_argument("--interactive", action="store_true", help="Allow Manus to ask follow-up questions")

    # status
    p_status = sub.add_parser("status", help="Check task status")
    p_status.add_argument("--task-id", required=True, help="Task ID")
    p_status.add_argument("--convert", action="store_true", help="Convert PPTX output when supported by Manus")

    # result
    p_result = sub.add_parser("result", help="Get task result and download attachments")
    p_result.add_argument("--task-id", required=True, help="Task ID")
    p_result.add_argument("--download-dir", default=None, help="Directory for downloads (default: ~/.manus-skill/downloads/YYYYMM/)")
    p_result.add_argument("--convert", action="store_true", help="Convert PPTX output when supported by Manus")

    # list
    p_list = sub.add_parser("list", help="List recent tasks")
    p_list.add_argument("--limit", type=int, default=10, help="Max tasks to return")
    p_list.add_argument("--status", default=None, help="Filter by status")

    # delete
    p_delete = sub.add_parser("delete", help="Delete a task via Manus API")
    p_delete.add_argument("--task-id", required=True, help="Task ID")

    # Backward-compatible alias for older skill docs. This deletes the task.
    p_cancel = sub.add_parser("cancel", help="Deprecated alias of delete")
    p_cancel.add_argument("--task-id", required=True, help="Task ID")

    args = parser.parse_args()

    handlers = {
        "create": cmd_create,
        "status": cmd_status,
        "result": cmd_result,
        "list": cmd_list,
        "delete": cmd_delete,
        "cancel": cmd_delete,
    }
    handlers[args.command](args)


if __name__ == "__main__":
    main()
